homeassistant/helpers/area_registry.py

```diff
@@ -133,14 +135,20 @@ def _data_to_save(self) -> dict:
 @bind_hass
 async def async_get_registry(hass: HomeAssistantType) -> AreaRegistry:
     """Return area registry instance."""
-    task = hass.data.get(DATA_REGISTRY)
+    reg = hass.data.get(DATA_REGISTRY)
+
+    if not reg:
+        evt = hass.data[DATA_REGISTRY] = Event()
 
-    if task is None:
-        async def _load_reg() -> AreaRegistry:
-            registry = AreaRegistry(hass)
-            await registry.async_load()
-            return registry
+        reg = AreaRegistry(hass)
+        await reg.async_load()
```

Comment:
If you're used to using threaded code this looks horribly un-threadsafe. Of course with AIO code the only place you can switch contexts is at this line, so everything above and below is atomic anyhow.